snewpdag/plugins/Chi2Calculation.py

```diff
@@ -0,0 +1,204 @@
+    def time_diff(self, det1, det2, n):
+        det1_pos = self.det_cartesian_position(det1)
+        det2_pos = self.det_cartesian_position(det2)
+        return float((det1_pos - det2_pos).getT() @ n)/3.0e8
+
+    # Calculates chi2 for given vector d
+    def chi2(self, d):
+        return (d.getT() @ (self.precision_matrix @ d))
+
+    # Calculates vector d given supernova position and t_m
+    def d_vec(self, n, dt_i):
+        n_detectors = len(self.detector_info)
+        d = [0]*n_detectors
+
+        for i in range(n_detectors):
+            det = list(self.detector_info.keys())[i]
+            det0 = self.d0_info[-1]
+            d[i] = dt_i[det][0] + dt_i[det][1] / 1e9 \
+                - dt_i[det0][0] - dt_i[det0][1] / 1e9
+
+            d[i] = d[i] - self.detector_info[det][4] + self.d0_info[4]
+
+            d[i] -= self.time_diff(self.detector_info[det], self.d0_info, n)
+
+        return np.matrix(d).getT()
+
+    # Estimates the SN direction,
+    # could be used for debugging purposes
+    def calculate_real_n(self, dt_i):
+        chi2_min = 1e12
+
+        for i in range(self.NPIX):
+            delta, alpha = hp.pixelfunc.pix2ang(self.NSIDE, i)
+            delta = delta - math.pi/2
+            alpha = alpha - math.pi
+
+            n_pointing = -1*self.angles_to_unit_vec(alpha, delta)
+            chi2_at_n = self.chi2(self.d_vec(n_pointing, dt_i))
+            if chi2_at_n < chi2_min:
+                chi2_min = chi2_at_n
+                n = n_pointing
+        return n
+
+    # Puts contours on a map
+    def draw_contours(self, map):
+        sigma1 = chi2.ppf(0.683, self.n_dof)
+        sigma2 = chi2.ppf(0.955, self.n_dof)
+        sigma3 = chi2.ppf(0.997, self.n_dof)
+
+        # Some arbitrary numbers that make good contrast
+        c1 = 10  # map.max()
+        c2 = 8  # (map.max() + map.min())/3
+        c3 = 7  # (map.max() + map.min())*2/3
+        out = 0  # map.min()
+
+        for i in range(len(map)):
+            if map[i] < sigma1:
+                map[i] = c1
+            elif map[i] < sigma2:
+                map[i] = c2
+            elif map[i] < sigma3:
+                map[i] = c3
+            else:
+                map[i] = out
+
+        return map
+
+    # Generates chi2 map
+    def generate_map(self, dt_i):
+        map = np.zeros(self.NPIX)
+
+        for i in range(self.NPIX):
+            delta, alpha = hp.pixelfunc.pix2ang(self.NSIDE, i)
+            delta = delta - math.pi/2
+            alpha = alpha - math.pi
+
+            n_pointing = -1*self.angles_to_unit_vec(alpha, delta)
+            map[i] = self.chi2(self.d_vec(n_pointing, dt_i))
+
+        return map
+
+    def alert(self, data):
+        dt_i = data['gen']['sn_times']
+        self.arrival = dt_i['Earth']
+
+        map = self.generate_map(dt_i)
+
+        if self.add_contours:
+            map = self.draw_contours(map)
+
+        hp.write_map(self.output_file, map, overwrite=True)
```

Comment:
I'd actually move the picture-drawing part to a new skymap renderer, called something like "ContourSkymap", along with draw_contours().  This plugin (which just does the calculation) should leave the healpix array in nested ordering (note, this isn't the default for healpy) in the payload.

Reply:
It may be good to leave the healpix array in a field named 'chi2', and store the number of degrees of freedom in 'ndof', so it can be combined with other measurements using CombineMaps.